rfc/rfc006-distributed-registry.md

```diff
@@ -0,0 +1,328 @@
+
+1. **Bob** generates his CA certificate (see [RFC008](rfc008-certificate-structure.md)) and creates the vendor proof (see section 6, "Vendor proof").
+2. **Bob** sends his CA certificate and vendor proof to **Alice** via an out-of-band mechanism (see below).
+3. **Alice** creates the vendor registration document with **Bob**'s CA certificate and vendor proof, then signs and publishes it.
+4. **Bob** adds **Alice**'s CA certificate to its truststore.
+5. **Bob** connects to **Alice** and starts receiving documents.
+
+Before **Alice** connects to **Bob** they communicate out-of-band. A (relatively) trusted means of communication SHOULD
+be used, like signed e-mail or authenticated chat.
+
+#### Fields
+The following fields MUST be present:
+* **certs** (certificates): MUST contain an array containing base64 encoded vendor CA certificates.
+* **prfs**: MUST contain an array of cryptographic proofs asserting the vendor being a real software company.
+
+Example:
+```json
+{
+    "certs": ["CA certificate as base64 ASN.1"],
+    "prfs": ["vendor proof"]
+}
+```
+
+#### Duplicates
+Vendors SHOULD take care not to register a duplicate organization, but if this happens the vendor and all related
+objects (organizations, services, endpoints, etc.) MUST be processed as long as they're valid. As such there can be
+multiple vendor registrations with the same vendor ID. 
+
+#### Validation
+When processing a vendor registration or update it MUST be validated as follows:
+
+1. Assert all required fields are present.
+2. Assert all fields value formats are valid.
+3. For updates: assert only modifiable fields are updated.
+4. Validate **prfs**:
+   * Assert there is at least one proof.
+   * Assert proofs are valid at the time they're introduced.
+   * Assert all proofs authenticate the same vendor.
+5. Validate **certs**:
+   * Assert there is at least one certificate.
+   * Assert all certificates conform to the vendor CA certificate specification (see [RFC008](rfc008-certificate-structure.md)).
+   * Assert all certificates have the same MUST contain the same vendor ID.
+
+If any of these steps fail the registration or update SHALL NOT be processed.
+
+#### Revoking CA certificates
+When a vendor loses or leaks its private key, the vendor SHOULD remove the certificate from its registration to avoid
+malicious parties exploiting it. However, other parties MUST regard signatures created before removal of the certificate
+ as valid. How the vendor SHOULD change its CA certificate depends on the situation:
+
+##### CA certificate private key loss
+In case the vendor loses its CA private key it MUST register a new vendor CA certificate. This can be done in two ways,
+depending on what was lost:
+
+1. If the vendor still has a valid signing certificate it COULD sign the update using that private key.
+2. Otherwise, when the signing certificate has expired or the private key is lost as well, the vendor MUST also provide
```

Comment:
And publish via another Vendor? Wouldn't it be easier to just add a new _registry/vendor_ doc?

Reply:
True, since the entries can just be merged (as mentioned in 'Duplicates')

Reply:
True, the ID should come from a third party (from certificate) which would it make to do an ID comparison?

Reply:
Added in spec. Note that **PKIOverheid Organisatie**-certificates don't contain a KVK-nummer (according to the KPN Helpdesk) so we still have a challenge to solve there...